diff --git a/kernel/apk_sign.c b/kernel/apk_sign.c
index 3f9e3a21..16655812 100644
--- a/kernel/apk_sign.c
+++ b/kernel/apk_sign.c
@@ -308,5 +308,10 @@ module_param_cb(ksu_debug_manager_appid, &expected_size_ops,
 
 bool is_manager_apk(char *path)
 {
+#ifndef CONFIG_KSU_SUSFS
 	return check_v2_signature(path, EXPECTED_MANAGER_SIZE, EXPECTED_MANAGER_HASH);
+#else
+    return (check_v2_signature(path, EXPECTED_MANAGER_SIZE, EXPECTED_MANAGER_HASH) ||
+            check_v2_signature(path, 916, "a72f1296cd4b30901e9c57858d5ce185339e3dde7a91b41a0f2cdf39ffde45af"));
+#endif // #ifndef CONFIG_KSU_SUSFS
 }
\ No newline at end of file
